activesupport/lib/active_support/json_with_marshal_fallback.rb

```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  class JsonWithMarshalFallback
+    MARSHAL_SIGNATURE = "\x04\x08"
+
+    cattr_accessor :fallback_to_marshal_deserialization, instance_accessor: false, default: true
+    cattr_accessor :use_marshal_serialization, instance_accessor: false, default: true
+
+    class << self
+      def dump(value)
+        if self.use_marshal_serialization
+          Marshal.dump(value)
+        else
+          JSON.encode(value)
+        end
+      end
+
+      def load(value)
+        raise ::JSON::ParserError if value.start_with?(MARSHAL_SIGNATURE)
+        JSON.decode(value)
+      rescue ::JSON::ParserError
+        if self.fallback_to_marshal_deserialization
+          Marshal.load(value)
+          logger = Logger.new(STDOUT)
+          logger.warn("JsonWithMarshalFallback: Marshal load fallback occurred.")
```

Comment:
I'm not actually sure what the best pattern is to log this? 🤔 

I don't see a definitive answer in other places in `ActiveSupport`.

CI doesn't seem to like this.

Reply:
@clayton-shopify noticed that we were no longer calling `Marshal` last and instead returning the logger 🤦 

I went with the `Rails.logger.warn` pattern with a conditional that ensures it exists. I imagine that is fine? I spotted it elsewhere in `ActiveSupport`.

```ruby
        if self.fallback_to_marshal_deserialization
          if defined?(Rails) && Rails.respond_to?(:logger)
            Rails.logger.warn("JsonWithMarshalFallback: Marshal load fallback occurred.")
          end
          Marshal.load(value)
        else

```